kadai1/shinji/convimg/convimg.go

```diff
@@ -0,0 +1,70 @@
+// 画像の変換機能を提供します。
+package convimg
+
+import (
+	"fmt"
+	"image"
+	"image/gif"
+	"image/jpeg"
+	_ "image/jpeg"
+	"image/png"
+	"os"
+	"path/filepath"
+)
+
+type Ext string
+
+const (
+	PNG  Ext = ".png"
+	JPEG Ext = ".jpg"
+	GIF  Ext = ".gif"
+)
+
+// 拡張子を変更します。
+func convExt(srcPath string, to Ext) string {
+	ext := filepath.Ext(srcPath)
+	return srcPath[:len(srcPath)-len(ext)] + string(to)
+}
+
+// エラー処理
+func assert(err error, msg string) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
+	}
+}
+
+// 画像を変換します。
+func Do(srcPath string, to Ext) {
+	// ファイルオープン
+	src, err := os.Open(srcPath)
```

Comment:
gosecを使うと、セキュアなコードが目指せるので、みてみると良きかもです！

> G304 (CWE-22): Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
  > > os.Open(srcPath)

```suggestion
	src, err := os.Open(filepath.Clean(srcPath))
```

Reply:
ありがとうございます！
gosecを使用して、セキュアなコードに修正しました。（Issuesを0にしました）

https://github.com/gopherdojo/gopherdojo-studyroom/blob/98cb6b2d75b6eba1feef7d3477c6b556962cf63b/kadai1/shinji/convimg/convimg.go#L55

一点質問させて下さい。
質問：ここでfilepath.Cleanを使用するとセキュアになるのはどうしてでしょうか？
（何のチェックもせずにファイルを開くと危険というのは何となくわかるのですが、filepath.Cleanでそれをどのように回避しているのかが今一つ分かりません）

試した事：
filepath.Cleanのドキュメント（https://golang.org/pkg/path/filepath/#Clean ）を読みました。
読んだ限りではパスを整形するもので、それがなぜセキュアに繋がるのかが分かりませんでした。

filepath.Cleanを使用した場合と使用しなかった場合のパスを標準出力してみたのですが、同じパスが表示され、違いが分かりませんでした。（今回はたまたまfilepath.Cleanの対象外だった？）

下記ページを読みましたが、やはりfilepath.Clean()でなぜセキュアに出来るのかは分かりませんでした。
https://securego.io/docs/rules/g304.html
https://kenzo0107.github.io/2019/12/08/2019-12-09-gosec-os-open/

以上、お時間のある時にご回答いただけますと幸いです。

Reply:
確認漏れてて遅くなりました、すみません！

こちら確かに、入れたパスと等価な最短のパスを生成する関数になります。
`Clean()`という関数名から、ファイルパスのサニタイズに使っているのかと勘違いしてました。。。申し訳ないです！🙏)

その上で調べてみたのですが、直接的にパスの整形がセキュアに繋がることはなさそうな感じがします。

あるとすれば、この関数によって綺麗な最短パスが返ってくるので、
カレントパスより上のディレクトリを指定されたとき(そして弾きたい)などに、
後からシンプルな正規表現で捉えて弾きやすいということなどかなと思います！

イメージこのような感じです！

```golang
cleaned := filepath.Clean(raw)
if strings.HasPrefix(cleaned, "../") {
  // アクセス禁止処理
}
```

[こちら](http://technosophos.com/2016/03/31/go-quickly-cleaning-filepaths.html)を参考にしました！

ただ実際のところHTTP経由のディレクトリトラバーサル攻撃は、`http.FileServer`で防げそう([参考](https://shogo82148.github.io/blog/2016/04/13/serving-static-files-in-golang/))なので、
これ必要？とは確かになりますね！

Reply:
いえ、ご回答くださりありがとうございます！
詳細に調べて下さりありがとうございます、大変参考になりました。
Clean()もhttp.FileServerも、場面に応じて使っていこうと思います！